File: 1026_MaximumDifferenceBetweenNodeAndAncestor/main.cpp
class Solution{
public:
    int maxAncestorDiff(TreeNode* root){
        return dfs(root, root->val, root->val);
    }
    
private:
    int dfs(TreeNode* root, int mn, int mx){
        if(root == NULL) return mx - mn;
        mn = min(mn, root->val);
        mx = max(mx, root->val);
        return max(dfs(root->left, mn, mx), dfs(root->right, mn, mx));
    }
}

File: 1026_MaximumDifferenceBetweenNodeAndAncestor/main.java
class Solution {
    public int maxAncestorDiff(TreeNode root) {
        return dfs(root, root.val, root.val);
    }
    
    private int dfs(TreeNode n, int mn, int mx){
        if(n == null) return mx - mn;
        mn = Math.min(mn, n.val);
        mx = Math.max(mx, n.val);
        return Math.max(dfs(n.left, mn, mx), dfs(n.right, mn, mx));
    }
}
